crates/sui-core/src/execution_scheduler/mod.rs

```diff
@@ -0,0 +1,1119 @@
+            let receiving_object_keys: HashSet<_> = tx_data
+                .receiving_objects()
+                .into_iter()
+                .map(|entry| {
+                    InputKey::VersionedObject {
+                        // TODO: Add support for receiving ConsensusV2 objects. For now this assumes fastpath.
+                        id: FullObjectID::new(entry.0, None),
+                        version: entry.1,
+                    }
+                })
+                .collect();
+            let input_and_receiving_keys = [
+                input_object_keys,
+                receiving_object_keys.iter().cloned().collect(),
+            ]
+            .concat();
+
+            let scheduler = self.clone();
+            let epoch = epoch_store.epoch();
+            let epoch_store = epoch_store.clone();
+            spawn_monitored_task!(
+                epoch_store.within_alive_epoch(scheduler.schedule_transaction(
+                    cert,
+                    expected_effects_digest,
+                    input_and_receiving_keys,
+                    receiving_object_keys,
+                    epoch,
+                ))
+            );
+        }
+
+        self.metrics
+            .transaction_manager_num_pending_certificates
+            .set(self.pending_certificates.read().len() as i64);
+    }
+
+    async fn schedule_transaction(
+        self,
+        cert: VerifiedExecutableTransaction,
+        expected_effects_digest: Option<TransactionEffectsDigest>,
+        input_and_receiving_keys: Vec<InputKey>,
+        receiving_object_keys: HashSet<InputKey>,
+        epoch: EpochId,
+    ) {
+        let digest = cert.digest();
+        let digests = [*digest];
+        tracing::debug!(?digest, "Schedule_transaction");
+        let enqueue_time = Instant::now();
+        tracing::trace!(
+            ?digests,
+            "Waiting for input objects: {:?}",
+            input_and_receiving_keys
+        );
+
+        tokio::select! {
+            _ = self.object_cache_read
+                .notify_read_input_objects(&input_and_receiving_keys, &receiving_object_keys, &epoch)
+                => {
+                    self.metrics
+                        .transaction_manager_transaction_queue_age_s
+                        .observe(enqueue_time.elapsed().as_secs_f64());
+                    tracing::debug!(?digests, "Input objects available");
+                    // TODO: Eventually we could fold execution_driver into the scheduler.
+                    let _ = self.tx_ready_certificates.send(PendingCertificate {
+                        certificate: cert,
+                        expected_effects_digest,
+                        stats: PendingCertificateStats {
+                            enqueue_time,
+                            ready_time: Instant::now(),
+                        },
+                    });
+                }
+            _ = self.transaction_cache_read.notify_read_executed_effects(&digests) => {
```

Comment:
this should just be `notify_read_executed_effects_digests` to avoid needlessly looking up the full effects